### PR TITLE
Smart homepage

### PR DESCRIPTION
**Pull request by burricks**
Automatically generates the sidebar menu items based on user permissions

### CODE REVIEW

File: ap/aputils/templatetags/smart_homepage.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+register = template.Library()
+
+#Helper Functions
+def my_reverse(url_pattern):
+	if url_pattern != '#':
+		return reverse(url_pattern)
+	else:
+		return '#'
+
+def smart_add(url, name):
```
Comment (radicalray) [inline ap/aputils/templatetags/smart_homepage.py]: better name: calculate_menu_item_url

File: ap/aputils/templatetags/smart_homepage.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		trainee_only=[SubMenuItem(name='View Schedule', url='schedules:schedule', condition=True), SubMenuItem(name='Personal Attendance', url='attendance:attendance-submit', condition=True), SubMenuItem(name='View Personal Leaveslips', url='leaveslips:leaveslips-list', condition=True)], 
+		common=[SubMenuItem(name='Submit Individual Leaveslips', url='leaveslips:individual-create'), SubMenuItem(name='Submit Group Leaveslips', url='leaveslips:group-create')], 
+		specific=[SubMenuItem(name='View Trainee Leaveslips', permission='attendance.add_roll', url='leaveslips:ta-leaveslip-list')])
+	discipline_menu = MenuItem(name ='Discipline', 
+		common =[SubMenuItem(name='Life Studies', url='lifestudies:discipline_list'), SubMenuItem(name='Class Notes', url='#')])
+	exam_menu = MenuItem(name = 'Exams',  
+		specific = [SubMenuItem(name="Create Exam", permission='exams.add_exam', url='exams:submit')])
+	requests_menu = MenuItem(name= 'Requests', 
+		common=[SubMenuItem(name='A/V Requests', url='#'), SubMenuItem(name='Maintenance Requests', url='#'), SubMenuItem(name='Room Reservations', url='#'), SubMenuItem(name='Web Access Requests', url='web_access:web_access-list')])
```
Comment (radicalray) [inline ap/aputils/templatetags/smart_homepage.py]: url_name

File: ap/aputils/templatetags/smart_homepage.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+def smart_add(url, name):
+	path = my_reverse(url)
+	return [(path, name)]
+
+
+#Generates the menu
+@register.assignment_tag(takes_context=True)
+def generate_menu(context):
+	menu = ""
+
```
Comment (radicalray) [inline ap/aputils/templatetags/smart_homepage.py]: menu = [
 {group_name: group, items: [ name1, name2]},

]

File: ap/ap/context_processors.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from django.db.models import Q
+
+def permissions(request):
+  user = request.user
+  return {
+    'exam.grade': user.groups.filter(Q(name='exam_graders') | Q(name='administration')).exists(),
+    'attendance.add_roll': user.groups.filter(Q(name='attendance_monitors') | Q(name='team_monitors') | Q(name='ypc_monitors')).exists(),
```
Comment (jontey) [inline ap/ap/context_processors.py]: Forgot to add HCs

**Comment by jontey** [comment]
Closing this as code is duplicated in #455

**Pull request closed without merging**